gobblin-utility/src/main/java/gobblin/util/AvroUtils.java

```diff
@@ -156,4 +170,116 @@ public static GenericRecord convertRecordSchema(GenericRecord record, Schema new
       closer.close();
     }
   }
+
+  /**
+   * Get the latest avro schema for a directory
+   * @param directory the input dir that contains avro files
+   * @param conf configuration
+   * @param latest true to return latest schema, false to return oldest schema
+   * @return the latest/oldest schema in the directory
+   * @throws IOException
+   */
+  public static Schema getDirectorySchema(Path directory, Configuration conf, boolean latest) throws IOException {
+    Schema schema = null;
+    Closer closer = Closer.create();
+    try {
+      List<FileStatus> files = getDirectorySchemaHelper(directory, FileSystem.get(conf));
+      FileStatus file = latest ? files.get(0) : files.get(files.size() - 1);
+      LOG.info("Path to get the avro schema: " + file);
+      FsInput fi = new FsInput(file.getPath(), conf);
+      GenericDatumReader<GenericRecord> genReader = new GenericDatumReader<GenericRecord>();
+      schema = new DataFileReader<GenericRecord>(fi, genReader).getSchema();
+    } catch (IOException e) {
+      throw new IOException("Cannot get the schema for directory " + directory);
+    } finally {
+      closer.close();
+    }
+    return schema;
+  }
+
+  private static List<FileStatus> getDirectorySchemaHelper(Path directory, FileSystem fs) throws IOException {
+    List<FileStatus> files = new ArrayList<FileStatus>();
+    if (fs.exists(directory)) {
+      getAllNestedAvroFiles(fs.getFileStatus(directory), files, fs);
+      if (files.size() > 0) {
+        Collections.sort(files, new Comparator<FileStatus>() {
+          @Override
+          public int compare(FileStatus file1, FileStatus file2) {
+            return Long.valueOf(file2.getModificationTime()).compareTo(Long.valueOf(file1.getModificationTime()));
+          }
+        });
+      }
+    }
+    return files;
+  }
+
+  private static void getAllNestedAvroFiles(FileStatus dir, List<FileStatus> files, FileSystem fs) throws IOException {
+    if (dir.isDir()) {
+      for (FileStatus f : fs.listStatus(dir.getPath())) {
+        getAllNestedAvroFiles(f, files, fs);
+      }
+    } else if (dir.getPath().getName().endsWith(AVRO_SUFFIX)) {
+      files.add(dir);
+    }
+  }
+
+  /**
+   * Merge oldSchema and newSchame. Set a field default value to null, if this field exists in the old schema but not in the new schema.
+   * @param oldSchema
+   * @param newSchema
+   * @return schema that contains all the fields in both old and new schema.
+   */
+  public static Schema nullifyFiledsForSchemaMerge(Schema oldSchema, Schema newSchema) {
+    if (oldSchema == null) {
+      return newSchema;
+    }
+
+    List<Field> combinedFields = new ArrayList<Field>();
+    for (Field newFld : newSchema.getFields()) {
+      combinedFields.add(new Field(newFld.name(), newFld.schema(), newFld.doc(), newFld.defaultValue()));
+    }
+
+    List<Schema> union = new ArrayList<Schema>();
```

Comment:
Yes, you are right. The error will be triggered when multiple fields are missing in the new schema. Fixed and new unit test added.